src/ql/ir/old_to_new.cc

```diff
@@ -229,14 +231,16 @@ static void parse_decomposition_rule(
             "expansions"
         );
     }
-    utils::UInt decomp_duration = get_duration_of_block(rule_ir->program->blocks[0]);
-    if (decomp_duration > ityp->duration) {
-        QL_USER_ERROR(
-            "in " << description.str() << ": the duration of the schedule of " <<
-            "the decomposition (" << decomp_duration << ") cannot be longer " <<
-            "than the duration of the to-be-decomposed instruction (" <<
-            ityp->duration << ")"
-        );
+    if (ityp->duration > 0) {   // 0 implies that durations are just added up, requires scheduling after decomposition
```

Comment:
I don't understand this.
Is allowing a duration of 0 a new convention for instruction definitions in the config with new-style-decompositions? I see that without this, you would get an error when the decomposition contains instructions with >0 duration. So, when you know that you would schedule anyhow, you don't want this error.
So, setting the duration to 0, avoids this error, nothing else.
What you would want to define with a particular decomposition is, that it must always be scheduled first, i.e. when the error condition applies, the cycle assignments in the block in which it gets expanded, become invalid.
So we are back at the cycles_invalid flag.

Reply:
Your analysis is correct, this allows disabling the check, allowing the decomposition to take place. It doesn't make sense to schedule an instruction with 0 duration, and decompose it afterwards.
The new IR - like the old one - can represent both unscheduled and scheduled programs. I don't know whether - and if so, where - this state is maintained or checked.

Reply:
I'm exploring how to address this, since setting the instruction duration to 0 is not general enough in the context of multiple decompositions, for different mapping strategies, for instance.
The documentation reads:

> the duration of the decomposed block may not be longer than the duration of the to-be-decomposed instruction (either make the instruction duration long enough, or define the decomposition as a single bundle and (re)schedule after applying the decomposition).

When you certainly will schedule after decomposition, you can as well specify the instruction's duration to be larger than you ever think the decomposition will take. I'd prefer that approach. It is also in line with the documentation.

To have a system of predicates around passes such as: `circuit-cycles-valid` is quite advanced, when you want to make it a general feature; other examples are: `has-been-decomposed-to-primitives` and `is-mapped-to-topology` which of course depend on the configuration/target.

Reply:
So I propose that you give the instruction a long duration and take the test on the duration being 0 out.

Reply:
With the 'predicates around passes', I guess you mean to somehow maintain in the IR in what state it is, and then act upon that. 
I'm not so sure whether it's possible to come up with a generic solution for that: a scheduler may set `circuit-cycles-valid`, but then again a later pass might invalidate that. So the semantics of these predicates would have to be globally agreed upon, and new passes will probably need new predicates, requiring updates on all other passes. 

Another approach, which is more or less the current approach, would be for every pass to check the IR on it's requirements, and fail if these are not met (e.g. if the CC backend encounters a non-native instruction that should be decomposed).

Anyway, this would be a discussion for future development, a bit out of scope for this PR

Reply:
My discussion of predicates around passes was more a reduction ad absurdum to come to the same conclusion as you do, that it is too far fetched now.
But I don't like the test on the duration being 0, or I don't understand the context of the current code, i.e. where this 0 duration comes from, from the config file or from somewhere internally in this old_to_new code.
I'm now trying to understand this.

Reply:
This is about adding legacy decompositions to instruction types in the new IR. Those legacy decomposition bodies (I'm guessing now) are kind of already expanded here when adding those to the instruction type. I would have expected this kind of error (that the expansion takes more cycles than the original instruction) to be given by the decomposition pass when actually performing the decomposition, and not here when making the rule. When done by the decomposition pass, it could use the pass value ignore_schedule in giving this error message, because when the schedule is ignored, the expansion can take more cycles than the original.
Anyhow, still studying, see my other conversation.

Reply:
Why did you add this test? A crash or convenience or was it a bug that it was absent?

Reply:
The test is part of function
```
/**
 * Parses a new-style decomposition rule.
 */
static void parse_decomposition_rule(
    const Ref &ir,
    const InstructionTypeLink &ityp,
    const utils::Json &json
) {
```
, a static helper for the *new-style* decompositions.

`ityp->duration` is taken from the `duration` field of the instruction in the JSON platform file.

When decompositions are intended to be used pre-scheduling, it is quite inconvenient to be forced to specify some duration. Specifying some 'large' number would work, but the trouble I have with that is that you then get into the situation where - depending on context - numbers are meaningful or not, and the context may be difficult to tell. A value that *explicitly* triggers special treatment is much more future-proof in my opinion, and 0 seems to be a safe value.

Reply:
*Comment by Hans that appeared lower down because Github's email interface messes up things*

What was your use case?
Why did you add this test?

So for you a sufficiently high value would work as well.
Or we take this test here out and put it at those places where the decomposition is actually done.
And use the ignore_schedule flag.

That an instruction can have a duration and a decomposition serves the purpose to treat it as a primitive until the decomposition is done.
E.g. scheduled, optimized, and mapped.

Reply:
The use case is basically any new style decomposition, but a maybe interesting example is this:

```
        "_do_measure": {
            "prototype": ["M:qubit"],
            "duration": 800,
            "cc": {
                "signal": [
                    {   "type": "measure",
                        "operand_idx": 0,
                        "value": ["dummy"] 
                    }
                ],
                "static_codeword_override": [0] 
            }
        },
        "measure": {
            "prototype": ["M:qubit"],
            "duration": 0,
            "decomposition": {
                "into": "_do_measure op(0)"
            }
        },
        "measure q6": {
            "prototype": ["M:qubit"],
            "duration": 0,
            "decomposition": {
                "into": [
                    "rx12 q[6];",
                    "_do_measure q[6]"
                ]
            }
        },
```

The purpose here is to treat measurements of some qubits (q6 in this case) differently, by prepending a gate (`rx12`, a rotation on the second excited state) to improve measurement fidelity. Note that these decompositions (need to) happen before scheduling.

The actual measurement ("_do_measure") has a duration (800) that is actually generated by a script where the user can configure durations of particular types of gates, so doing the math by hand (800 for "measure" and 820 for "measure q6") is impractical and error-prone. The two decompositions of "measure" use the change discussed here to allow a duration of 0 to mean 'ignore'.

Of course is would be *possible* to use a large number (1000?, 1000000?), but these numbers tend to come with assumptions that may not hold over time, so I think it's bad design to require that. And I don't think that use of the value 0 for special treatment could ever conflict with the wish for a decomposition to actually take zero time.

But I guess we just disagree on what's the best solution